File: Pod/Classes/ABEImgurAPIClient.m
//
//  ABEImgurAPIClient.m
//  IssueReporter
//
//  Created by Nikolai Johan Heum on 04.05.15.
//  Copyright (c) 2015 Abello. All rights reserved.
//

#import "ABEImgurAPIClient.h"
#import "ABEReporter.h"

@interface ABEImgurAPIClient ()

@property (nonatomic) NSMutableURLRequest *baseImageUploadRequest;

@end

@implementation ABEImgurAPIClient

+ (ABEImgurAPIClient *)sharedClient {
    
    static ABEImgurAPIClient *_sharedClient = nil;
    static dispatch_once_t onceToken;
    
    dispatch_once(&onceToken, ^{
        _sharedClient = [[self alloc] init];
        
    });
    
    return _sharedClient;
}

+ (BOOL)isAPIKeySet {
    return [self sharedClient].imgurAPIKey.length > 0;
}

- (void)uploadImageData:(NSData *)imageData success:(void (^)(NSString *imageURL))success error:(void (^)(NSError *))errorHandler {
    
    if (![ABEImgurAPIClient isAPIKeySet]) {
        // TODO : Refactor out error
        NSDictionary *userInfo = @{NSLocalizedDescriptionKey : @"Authentication required for Imgur",
                                   NSLocalizedRecoverySuggestionErrorKey : @"Did you set the imgur client key in your appdelegate using `setupWithRepositoryName:gitHubAccessToken:imgurClientID:`? If you do not have a key, go to https://api.imgur.com/oauth2/addclient and create one. This error has also been printed to the console."};
        
        errorHandler([NSError errorWithDomain:@"no.abello.IssueReporter" code:NSURLErrorUserAuthenticationRequired userInfo:userInfo]);
        return;
    }
    
    NSURLRequest *request = [self imageUploadRequestForImageData:imageData errorHandler:errorHandler];
    
    if (!request) {
        return;
    }

    [[[NSURLSession sharedSession] dataTaskWithRequest:request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
        if (error) {
            NSLog(@"There was an error while uploading the picture with Imgur's API.");
            NSLog(@"Error : %@", error);
            errorHandler(error);
            return;
        }
        
        NSError *jsonError;
        NSDictionary *jsonResponse = [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
        
        if (error) {
            NSLog(@"There was an error parsing the returned data from Imgur's API.");
            NSLog(@"Error : %@", jsonError);
        }
        
        NSString *urlString = jsonResponse[@"data"][@"link"];
        
        if (![NSURL URLWithString:urlString]) {
            NSLog(@"There was an error getting the url of the uploaded image from the response body of the Imgur API.");
            NSLog(@"Returned URL : %@", urlString);
            errorHandler(nil);
            return;
        }
        
        success(urlString);
    }] resume];
}

- (NSURLRequest *)imageUploadRequestForImageData:(NSData *)imageData errorHandler:(void (^)(NSError *))errorHandler {
    NSDictionary *parameters = @{@"image" : [imageData base64EncodedStringWithOptions:kNilOptions],
                                 @"type" : @"base64"};
    
    NSAssert([NSJSONSerialization isValidJSONObject:parameters], @"JSON Post body generated from issue is not valid JSON.");
    
    NSMutableURLRequest *baseRequest = [[self baseImageUploadRequest] copy];
    
    NSError *error = nil;
    NSData *data = [NSJSONSerialization dataWithJSONObject:parameters options:kNilOptions error:&error];
    
    if (error) {
        NSLog(@"There was an error serializing the upload image request body.");
        NSLog(@"Error : %@", error);
        errorHandler(error);
        return nil;
    }
    
    [baseRequest setValue:[NSString stringWithFormat:@"%ld", (unsigned long)[data length]] forHTTPHeaderField:@"Content-Length"];
    [baseRequest setHTTPBody:data];
    
    return baseRequest;
}

- (NSMutableURLRequest *)baseImageUploadRequest {
    if (!_baseImageUploadRequest) {
        NSURL *url = [NSURL URLWithString:@"https://api.imgur.com/3/upload"];
        _baseImageUploadRequest = [[NSMutableURLRequest alloc] initWithURL:url];
        
        _baseImageUploadRequest.HTTPMethod = @"POST";
        
        [_baseImageUploadRequest setValue:@"application/json" forHTTPHeaderField:@"Accept"];
        [_baseImageUploadRequest setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
        
        [_baseImageUploadRequest setValue:[NSString stringWithFormat:@"Client-ID %@", self.imgurAPIKey] forHTTPHeaderField:@"Authorization"];
    }
    
    return _baseImageUploadRequest;
}

- (void)setImgurAPIKey:(NSString *)imgurAPIKey {
    _imgurAPIKey = [imgurAPIKey copy];
}

@end

File: Pod/Classes/ABEImgurAPIClient.swift
//
//  ABEImgurAPIClient.swift
//  Pods
//
//  Created by Hakon Hanesand on 10/8/16.
//
//

import Foundation
import CFNetwork

enum IssueReporterError: Error {
    
    static let domain = "IssueReporter.IssueReporterError"
    
    case missingInformation(name: String)
    case invalid(name: String)
    
    case missingURLKeyInJSON
    
    case invalidURL
    case malformedResponseURL
    case unparseableResponse
    
    case jsonError(underlyingError: Error)
    case network(response: HTTPURLResponse, detail: String?)
    
    case error(error: Error?)
    
    
    
    var message : String {
        switch self {
        case let .missingInformation(name):
            return "The \(name) was not provided. Consult the README.md for information on how to acquire it."
        case .missingURLKeyInJSON:
            return "The uploaded image link was not present in the returned json."
            
        case .invalidURL:
            return "There was an error constructing the request URL."
        case let .invalid(name):
            return "Your \(name) is invalid, please follow the instructions in README.md"
        case .malformedResponseURL:
            return "There was an error extracting the image link from the returned response."
            
        case let .jsonError(underlyingError):
            return "JSON Error, underlying error : \(underlyingError)"
            
        case let .network(response, detail):
            let detail = detail != nil ? " with detail : " + detail! : ""
            return  "Network error with status code \(response.statusCode) "  + detail
            
        case let .error(error):
            return error == nil ? "Error : \(error)" : ""
            
        default:
            return ""
        }
    }
    
    var localizedDescription: String {
        return message
    }
}

final class ABEImgurAPIClient {
    
    public static var imgurAPIKey: String? = nil
    
    public static let shared = ABEImgurAPIClient()
    
    private init() { }
    
    func baseImageUploadRequest() throws -> URLRequest {
        
        guard let imgurAPIKey = ABEImgurAPIClient.imgurAPIKey else {
            throw IssueReporterError.missingInformation(name: "imgur api key")
        }
        
        guard let url = URL(string: "https://api.imgur.com/3/upload") else {
            throw IssueReporterError.invalidURL
        }
        
        var request = URLRequest(url: url)
        
        request.httpMethod = "POST"
        
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Client-ID \(imgurAPIKey)", forHTTPHeaderField: "Authorization")
        
        return request
    }
    
    fileprivate func uploadRequestForImageData(imageData: Data) throws -> URLRequest {
        let parameters = ["image" : imageData.base64EncodedString(),
                          "type" : "base64"]
        
        var baseIssueRequest = try baseImageUploadRequest()
        let data = try JSONSerialization.data(withJSONObject: parameters)
        
        baseIssueRequest.httpBody = data
        baseIssueRequest.setValue("\(data.count)", forHTTPHeaderField: "Content-Length")
        
        return baseIssueRequest
    }
    
    public func upload(imageData: Data, dispatchQueue: DispatchQueue = DispatchQueue.main, errorHandler: @escaping (IssueReporterError) -> (), success: @escaping (URL) -> ()) throws {
        
        let request = try self.uploadRequestForImageData(imageData: imageData)
        
        URLSession.shared.dataTask(with: request) { (data, response, error) in
            do {
                if let error = error {
                    throw IssueReporterError.error(error: error)
                }
                
                guard let response = response as? HTTPURLResponse, let data = data else {
                    throw IssueReporterError.unparseableResponse
                }
                
                if response.statusCode == 403 {
                    throw IssueReporterError.invalid(name: "client id")
                }
                
                let json = try JSONSerialization.jsonObject(with: data, options: []) as! NSDictionary
                
                guard let linkString = json.value(forKeyPath: "data.link") as? String else {
                    throw IssueReporterError.network(response: response, detail: json.value(forKeyPath: "data.error") as? String)
                }
                
                guard let url = URL(string: linkString) else {
                    throw IssueReporterError.malformedResponseURL
                }
                
                dispatchQueue.async {
                    success(url)
                }
            } catch let error as NSError where error.domain != IssueReporterError.domain {
                // Catch error not from our domain and wrap them
                errorHandler(IssueReporterError.jsonError(underlyingError: error))
            } catch {
                // Catch and forward all other errors
                errorHandler(error as! IssueReporterError)
            }
        }.resume()
    }
}
